src/Http/Http.Extensions/src/HttpResponseJsonExtensions.cs

```diff
@@ -185,7 +190,9 @@ public static Task WriteAsJsonAsync(
         response.ContentType = contentType ?? ContentTypeConstants.JsonContentTypeWithCharset;
 
         var startTask = Task.CompletedTask;
-        if (!response.HasStarted)
+        // Don't call StartAsync for IAsyncEnumerable. Headers might be set at the beginning of the generator which isn't invoked until
+        // JsonSerializer starts iterating over the IAsyncEnumerable.
+        if (!response.HasStarted && value is not null && !AsyncEnumerableHelper.IsIAsyncEnumerable(value.GetType()))
```

Comment:
Can we add a method to AsyncEnumerableHelper like the following?

```csharp
Task StartResponseIfPossible(this HttpResponse response, Type? typeToSerialize)
```

It seems like it would reduce a lot of repetitive `HasStarted` and null checks.